File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
private:
    void DFS(string cur, vector<string>& ans, int open, int close, int& n) {
        if(open > n || close > n || close > open) return;
        if(close == n) {
            ans.push_back(cur);
            return;
        }
        DFS(cur + "(", ans, open + 1, close, n);
        DFS(cur + ")", ans, open, close + 1, n);
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string cur = "";
        DFS(cur + "(", ans, 1, 0, n);
        return ans;
    }
};

File: 22-generate-parentheses/22-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        process(res, "", 0, 0, n);
        return res;
    }

    public void process(List<String> res, String cur, int open, int close, int n){
        if(cur.length() == 2 * n){
            res.add(cur);
            return;
        }

        if(open < n){
            process(res, cur + "(", open + 1, close, n);
        }

        if(close < open){
            process(res, cur + ")", open, close + 1, n);
        }
        return;

    }
}